ksql-engine/src/main/java/io/confluent/ksql/engine/TopicAccessValidator.java

```diff
@@ -0,0 +1,148 @@
+    for (String kafkaTopic : sourceTopicsExtractor.getSourceTopics()) {
+      checkAccess(kafkaTopic, AclOperation.READ);
+    }
+  }
+
+  private void validateCreateAsSelect(final CreateAsSelect createAsSelect) {
+    /*
+     * Check topic access for CREATE STREAM/TABLE AS SELECT statements.
+     *
+     * Validates Write on the target topic if exists, and Read on the query sources topics.
+     *
+     * The Create access is validated by the TopicCreateInjector which will attempt to create
+     * the target topic using the same ServiceContext used for validation.
+     */
+
+    validateQueryTopicSources(createAsSelect.getQuery());
+
+    // At this point, the topic should have been created by the TopicCreateInjector
+    final String kafkaTopic = getSourceTopicFromCreateAsSelect(createAsSelect);
+    checkAccess(kafkaTopic, AclOperation.WRITE);
+  }
+
+  private void validateInsertInto(final InsertInto insertInto) {
+    /*
+     * Check topic access for INSERT INTO statements.
+     *
+     * Validates Write on the target topic, and Read on the query sources topics.
+     */
+
+    validateQueryTopicSources(insertInto.getQuery());
+
+    final String kafkaTopic = getSourceTopicName(insertInto.getTarget().getSuffix());
+    checkAccess(kafkaTopic, AclOperation.WRITE);
+  }
+
+  private String getSourceTopicName(final String streamOrTable) {
+    final DataSource<?> dataSource = metaStore.getSource(streamOrTable);
+    if (dataSource == null) {
+      throw new KsqlException("Cannot validate for topic access from an unknown stream/table: "
+          + streamOrTable);
+    }
+
+    return dataSource.getKafkaTopicName();
+  }
+
+  /**
+   * Checks if the ServiceContext has access to the topic with the specified AclOperation.
+   */
+  private void checkAccess(final String topicName, final AclOperation operation) {
+    final Set<AclOperation> authorizedOperations = serviceContext.getTopicClient()
+        .describeTopic(topicName).authorizedOperations();
+
+    if (!authorizedOperations.contains(operation)) {
+      // This error message is similar to what Kafka throws when it cannot access the topic
+      // due to an authorization error. I used this message to keep a consistent message.
+      throw new KsqlException(String.format(
+              "Failed to %s Kafka topic: [%s]%n"
+                  + "Caused by: Not authorized to access topic",
+              StringUtils.capitalize(operation.toString().toLowerCase()), topicName)
+      );
+    }
+  }
+
+  private String getSourceTopicFromCreateAsSelect(final CreateAsSelect createAsSelect) {
+    // Get the topic if WITH(kafka_topic='') clause is used
+    final Expression nameExpression = createAsSelect.getProperties()
+        .get(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY);
+
+    if (nameExpression != null) {
+      return StringUtils.strip(nameExpression.toString(), "'");
+    }
+
+    return getSourceTopicName(createAsSelect.getName().getSuffix());
```

Comment:
I don't think we ever hit this line, because we always set`KAFKA_TOPIC_NAME_PROPERTY`. If we do hit this line I think it will throw an exception.

Reply:
It fails with the `TopicAccessValidatorTest` if I remove it. Perhaps the Server already manages to add the topic in the `KAFKA_TOPIC_NAME_PROPERTY`, but unit tests don't work unless I do mimic the behavior to set the property too.

I saw this the `TopicCreateInjector` was getting the info from the Metastore, so I did it when working on this class. I recalled this was called when the check was done in the `EngineExecutor`. 

Can I leave it for now until I write all the functional tests? I can go back to this line and see if it is really necessary later, and remove it as a bugfix.